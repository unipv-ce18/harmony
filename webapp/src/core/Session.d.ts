export class Session {
    getAccessToken(): Promise<string>
}
