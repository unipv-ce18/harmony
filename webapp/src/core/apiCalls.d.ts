import {ThemeId} from '../components/theme';

declare namespace apiCalls {

    export type AccessToken = string;
    type RefreshToken = string;

    type UserId = string;
    type ArtistId = string;
    type ReleaseId = string;
    type SongId = string;
    type PlaylistId = string;

    type LoginResult = {
        access_token: AccessToken,
        refresh_token: RefreshToken,
        token_type: 'bearer',
        expires_in: number,
        refresh_expires_in: number
    };

    type RefreshResult = {
        access_token: AccessToken,
        expires_in: number
    }

    type UserType = 'basic' | 'creator';
    type UserTier = 'free' | 'pro';
    type UserPreferences = {
        private: {email: boolean},
        theme: ThemeId
    }

    type UserData = {
        id: UserId,
        username: string,
        email: string,
        avatar_url?: string,
        bio: string,
        type: UserType,
        tier: UserTier,
        prefs: UserPreferences,
        artists?: any[]
    };

    type UploadContentCategory = 'user' | 'artist' | 'release' | 'song';
    type UploadContentObjectId = UserId | ArtistId | ReleaseId | SongId;
    type UploadContentResult = [string, any];  // Url and presigned post data

    // TODO: Update to the new API
    type SearchType = 'any' | 'artists' | 'releases' | 'songs' | 'playlists';

    type SearchResult = {
      artists?: [{id: ArtistId, name: string /* ...more */}]
      releases?: [{id: ReleaseId, name: string /* ...more */}]
      songs?: [{id: SongId, title: string /* ...more */}]
      playlists?: [{id: PlaylistId, /* ...more */}]
    }

    type ArtistResult = {
      id: ArtistId,
      name: string,
      creator: string | null,
      sort_name: string,
      country?: string
      life_span?: {begin: string, end: string | null},
      genres?: string[],
      bio?: string | null,
      members?: [{role: string, name: string}],
      links?: {[target: string]: string},
      image?: string | null,
      releases?: ReleaseResult[]
    }

    type ReleaseResult = {
      id: ReleaseId,
      name: string,
      date?: string,
      artist: any,  // keep it simple for now
      type: 'album' | 'single' | 'ep' | 'compilation' | 'live' | 'remix' | null,
      cover: string | null
    }

    function execLogin(identity: string, password: string): Promise<LoginResult>;

    function execLogout(token: AccessToken): Promise<void>;

    function execRefresh(refreshToken: RefreshToken): Promise<RefreshResult>;

    function getUserInfo(token: AccessToken, userId: UserId, includeArtists?: boolean): Promise<UserData>;

    function upgradeUserType(token: AccessToken): Promise<void>;

    function upgradeUserTier(token: AccessToken): Promise<void>;

    function patchUser(token: AccessToken, userId: UserId, patch: Partial<UserData>): Promise<void>;

    function uploadContent(category: UploadContentCategory, categoryId: UploadContentObjectId,
                           mimeType: string, size: number, token: AccessToken): Promise<UploadContentResult>;

    function uploadToStorage(result: UploadContentResult, file: File): Promise<Response>;

    function execSearch(token: AccessToken, type: SearchType, query: string): Promise<SearchResult>;

    function getArtist(artistId: ArtistId, withReleases: boolean, token: AccessToken): Promise<ArtistResult>;
}

export = apiCalls;
