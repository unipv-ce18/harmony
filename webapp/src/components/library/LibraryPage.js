import {Component} from 'preact';

import styles from './LibraryPage.scss';

import {catalog} from '../../Harmony';
import SongsPage from './SongsPage';
import PlaylistsPage from './PlaylistsPage';
import ArtistsPage from './ArtistsPage';
import ReleasesPage from './ReleasesPage';

class LibraryPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: "playlists",
      library: {},
      valid: false
    };

    this.showMenu = this.showMenu.bind(this);
  }

  showMenu(type) {
    this.setState({show: type});
  }

  componentDidMount() {
    catalog.getFullLibrary(/[^/]*$/.exec(window.location.href)[0])
      .then(result => {
        this.setState({library: result});
        this.setState({valid: true});
      });
  }

  render() {
    return (
      <div class={styles.libraryPage}>
        <div class={styles.libraryPageContent}>
          <div>
            <h1>YOUR LIBRARY</h1>
            <div>
              <div class={styles.inputs}>
                <span>
                  <button onClick={()=>this.showMenu("playlists")}>PLAYLISTS</button>
                  <button onClick={()=>this.showMenu("artists")}>ARTISTS</button>
                  <button onClick={()=>this.showMenu("releases")}>RELEASES </button>
                  <button onClick={()=>this.showMenu("songs")}>SONGS </button>
                </span>
              </div>
                {this.state.valid && this.state.show === "playlists" && <PlaylistsPage playlists={this.state.library.playlists}/>}
                {this.state.valid && this.state.show === "artists" && <ArtistsPage artists={this.state.library.artists}/>}
                {this.state.valid && this.state.show === "releases" && <ReleasesPage releases={this.state.library.releases}/>}
                {this.state.valid && this.state.show === "songs" && <SongsPage songs={this.state.library.songs}/>}
            </div>
          </div>
        </div>
      </div>);
  }
}

export default LibraryPage;
