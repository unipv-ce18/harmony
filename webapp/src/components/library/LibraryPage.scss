@import "../../styledefs";

.libraryPage {
  width: 100%;
  padding: 16px;
  box-sizing: border-box;
}

.libraryPageContent {
  width: 100%;
  border-radius: 16px;
  min-height: calc(100vh - 32px - #{$navbar-height});

  background-color: $panel-color-test;
  //background: radial-gradient(circle, #5B1A27 20%,#2A0C12 75%);

  color: $panel-text-color-test;
  overflow: hidden;
  display: inline-block;
  text-align: center;
  padding: 20px 0;
}

.libraryCommon {
  margin: 50px auto;
  div {
    div {
      color: $panel-color-test;
      background: linear-gradient(to right,
            rgba(255,255,255, 0),
            white,
            rgba(255,255,255, 0)
          );
    }

    span {
      display: inline-block;
      margin: 20px 8%;
      text-align: left;

      img {
        width: 150px;
        height: 150px;
      }

      p {
        margin: 0;

        a {
          padding: 0;
          margin: 0;
          background: transparent;
          border: none;
          cursor: pointer;
          color: white;
        }
      }
    }
  }
}

.libraryArtists {
  margin: 50px auto;
  span {
    display: inline-block;
    margin: 20px 8%;
    text-align: center;

    img {
      width: 150px;
      height: 150px;
      border-radius: 20px 0 20px 0;
    }

    p {
      margin: 0;

      a {
        padding: 0;
        margin: 0;
        background: transparent;
        border: none;
        cursor: pointer;
        color: white;
      }
    }
  }
}

.librarySongs table {
  margin: 50px auto;
  text-align: left;
  border-collapse: collapse;

  td, th { padding: 10px 30px }
  td:first-child { border-radius: 15px 0 0 15px }
  td:last-child { border-radius: 0 15px 15px 0 }
  tr:not(:first-child):hover td { background: black }

  th {
    text-align: center;
    padding-bottom: 30px;

    button {
      padding: 0;
      margin: 0;
      background: transparent;
      border: none;
      cursor: pointer;
      color: white;
      text-decoration: underline;
    }
  }
}


.inputs {
  width: 90%;
  margin-left: 5%;
  button{
    width: 200px;
  }
}
