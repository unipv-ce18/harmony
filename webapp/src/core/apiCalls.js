const API_LOGIN_URL = API_BASE_URL + '/auth/login';
const API_REGISTRATION_URL = API_BASE_URL + '/auth/register';
const API_LOGOUT_URL = API_BASE_URL + '/auth/logout';
const API_SEARCH_URL = API_BASE_URL + '/search';
const API_REFRESH_URL = API_BASE_URL + '/auth/refresh';
const API_RELEASE_URL = API_BASE_URL + '/release';
const API_PLAYLIST_URL = API_BASE_URL + '/playlist';
const API_ARTIST_URL = API_BASE_URL + '/artist';
const API_SONG_URL = API_BASE_URL + '/song';
const API_USER_URL = API_BASE_URL + '/user/';

export class ApiError extends Error {
  constructor(response) {
    super();
    this.name = 'ApiError';
    this.response = response;
  }
}

export function execLogin(identity, password) {
  const data = {identity, password};
  return fetch(API_LOGIN_URL, {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json'}),
    body: JSON.stringify(data)
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function execRegistration(email, username, password) {
  const data = {email, username, password};
  return fetch(API_REGISTRATION_URL, {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json'}),
    body: JSON.stringify(data)
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function execLogout(token) {
  return fetch(API_LOGOUT_URL, {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token})
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return true;
  });
}

export function execRefresh(token) {
  return fetch(API_REFRESH_URL, {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token}),
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function execSearch(token, type, query) {
  return fetch(`${API_SEARCH_URL}?q=${query}&t=${type}`, {
    headers: new Headers({'Authorization':'Bearer ' + token}),
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function getUserInfo(token, user_id, include_artists = false) {
  let query = API_USER_URL + user_id;
  if (include_artists) query += '?artists=1';
  return fetch(query, {
    method: 'GET',
    headers: new Headers({'Accept': 'application/json', 'Authorization':'Bearer ' + token})
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function changeUserType(token) {
  return fetch(API_USER_URL + 'type', {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token})
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return;
  });
}

export function changeUserTier(token) {
  return fetch(API_USER_URL + 'tier', {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token})
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return;
  });
}

export function patchUser(token, user_id, bio, prefs) {
  const data = {bio, prefs};
  return fetch(API_USER_URL + user_id, {
    method: 'PATCH',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token}),
    body: JSON.stringify(data)
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return;
  });
}

export function getReleasePlaylist(type_collection, collection_id, with_song, token) {
  // with_song == true includes the songs in the result
  let query;
  type_collection==='release'
    ? query = API_RELEASE_URL + '/' + collection_id
    : query = API_PLAYLIST_URL + '/' + collection_id;
  if (with_song) query += '?songs=1';
  return fetch(query, {
    method: 'GET',
    headers: new Headers({'Accept': 'application/json', 'Authorization':'Bearer ' + token}),
  }).then(response => {
      if (!response.ok) throw new ApiError(response);
      return response.json()
    })
}

export function getArtist(artist_id, with_releases, token) {
  // with_releases == true includes the songs in the result
  let query = API_ARTIST_URL + '/' + artist_id;
  if (with_releases) query += '?releases=1';
  return fetch(query, {
    method: 'GET',
    headers: new Headers({'Accept': 'application/json', 'Authorization':'Bearer ' + token}),
  }).then(response => {
      if (!response.ok) throw new ApiError(response);
      return response.json()
    })
}

export function createArtist(name, token) {
  const data = {name};
  return fetch(API_ARTIST_URL, {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token}),
    body: JSON.stringify(data)
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function deleteArtist(artist_id, token) {
  return fetch(API_ARTIST_URL + '/' + artist_id, {
    method: 'DELETE',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token})
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return;
  });
}

export function createRelease(artist_id, name, token) {
  const data = {artist_id, name};
  return fetch(API_RELEASE_URL, {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token}),
    body: JSON.stringify(data)
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function deleteRelease(release_id, token) {
  return fetch(API_RELEASE_URL + '/' + release_id, {
    method: 'DELETE',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token})
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return;
  });
}

export function deleteSong(song_id, token) {
  return fetch(API_SONG_URL + '/' + song_id, {
    method: 'DELETE',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token})
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return;
  });
}

export function getLibrary(user_id, token, full) {
  let query = API_USER_URL + user_id + '/library'
  if (full) query += '?full=1';
  return fetch(query, {
    method: 'GET',
    headers: new Headers({'Accept': 'application/json', 'Authorization':'Bearer ' + token}),
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function setLike(function_type, token, media_type, media_id) {
  const data = {media_type, media_id};
  return fetch(API_USER_URL + 'library', {
    method: function_type,
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token}),
    body: JSON.stringify(data)
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return true;
  });
}

export function getUserPlaylists(token) {
  return fetch(API_USER_URL + 'playlist', {
    method: 'GET',
    headers: new Headers({'Accept': 'application/json', 'Authorization':'Bearer ' + token})
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function createPlaylist(name, token) {
  const data = {name};
  return fetch(API_PLAYLIST_URL, {
    method: 'POST',
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token}),
    body: JSON.stringify(data)
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return response.json();
  });
}

export function updateSongInPlaylist(type_method, playlist_id, song_id, token) {
  let data;
  if (song_id !== null) data = {song_id};
  return fetch(API_PLAYLIST_URL + '/' + playlist_id, {
    method: type_method,
    headers: new Headers({'Content-Type': 'application/json', 'Authorization':'Bearer ' + token}),
    body: JSON.stringify(data)
  }).then(response => {
    if (!response.ok) throw new ApiError(response);
    return 'song added successfully';
  });
}
