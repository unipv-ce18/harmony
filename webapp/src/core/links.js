import {MediaItemInfo} from '../player/MediaPlayer';

export const artistLink = id => `/artist/${id}`;

export const releaseLink = id => `/release/${id}`;

// Not really a link but allows "linking" to tke player given a song model
export function createMediaItemInfo(song) {
  return new MediaItemInfo(song.id, {
    [MediaItemInfo.TAG_TITLE]: song.title,
    [MediaItemInfo.TAG_RELEASE]: song.release.name,
    [MediaItemInfo.TAG_ARTIST]: song.artist.name,
    [MediaItemInfo.TAG_ALBUMART_URL]: song.release.cover
  });
}
