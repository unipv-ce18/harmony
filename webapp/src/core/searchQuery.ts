export type SearchQuery = string;
type SearchModifier = {key: string, value?: string};
type QueryParams = {text: string, modifiers: SearchModifier[]};

const inputToUrl = (text: string) => text.replace(/[^a-z0-9]/gi, ' ').replace(/ /g, '+');
const inputFromUrl = (text: string) => text.replace(/\+/g, ' ');

export const isQueryEmpty = ({text, modifiers}: QueryParams) =>
  text === '' && modifiers.length === 0;

export function toSearchQuery(text: string, modifiers: SearchModifier[]): SearchQuery {
  const modStrings = modifiers.map(m => m.key + (m.value != null ? `=${inputToUrl(m.value)}` : ''));
  return inputToUrl(text) + (modStrings.length > 0 ? `:${modStrings.join(',')}` : '');
}

export function fromSearchQuery(data: SearchQuery): QueryParams {
  const [text, modStr] = data.split(':');

  const modifiers = modStr != null ? modStr.split(',').map(s => {
    const [key, value] = s.split('=');
    return {key, ...(value && {value: inputFromUrl(value)})};
  }) : []

  return {text: inputFromUrl(text), modifiers};
}
