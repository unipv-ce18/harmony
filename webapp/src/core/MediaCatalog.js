import {execLogin, execRefresh, getLibrary, setLike} from "./apiCalls";
import library from '../components/library/testLibrary';
import {session} from '../Harmony';


const LIBRARY_STORE_KEY = 'library';

export class MediaCatalog {

  #libraryCache = undefined;

  constructor(session) {
    this.session = session;
  }

  get #store() {
    if (this.#libraryCache === undefined)
      this.#libraryCache = JSON.parse(window.localStorage.getItem(LIBRARY_STORE_KEY));
    return this.#libraryCache;
  }

  set #store(value) {
    this.#libraryCache = value;
    if (value == null) {
      window.localStorage.removeItem(LIBRARY_STORE_KEY);
    }
    else {
      window.localStorage.setItem(LIBRARY_STORE_KEY, JSON.stringify(value));
    }
  }

  /**
   * Attempts to perform login with the API
   *
   * This will fail if the user is already logged in (i.e. loggedIn is true), credentials are invalid
   * or a server error occurs.
   *
   */
  setCachedLibrary() {
    let token = this.session.getAccessToken();
    return getLibrary('me', token, false)
      .then(data => {
        this.#store = data;
      });
  }

  getFullLibrary(user_id) {
    let token = this.session.getAccessToken();
    console.log("getFullLibrary")
    console.log(token)
    return getLibrary(user_id, token, true)
      .catch(e => console.log(e));
  }

  inLibrary(media_type, element_id) {
    if (media_type === 'songs') return this.#store.songs.includes(element_id);
    if (media_type === 'artists') return this.#store.artists.includes(element_id);
    if (media_type === 'releases') {
      return this.#store.releases.includes(element_id);
    }
    return false;
  }

  favorite(function_type, media_type, element_id) {
    let elem, bool = false;
    if (media_type === 'songs') elem = this.#libraryCache.songs;
    if (media_type === 'artists') elem = this.#libraryCache.artists;
    if (media_type === 'releases') elem = this.#libraryCache.releases;
    if (function_type === 'PUT' && !this.inLibrary(media_type, element_id)) {
      elem.push(element_id);
      bool = true;
    }
    if (function_type === 'DELETE' && this.inLibrary(media_type, element_id)) {
      const index = elem.indexOf(element_id);
      if (index > -1) elem.splice(index, 1);
      bool = true;
    }
    if (bool) {
      this.#store = this.#libraryCache;
      setLike(function_type, session.getAccessToken(), media_type, element_id)
        .catch(e => console.log(e));
    }
  }
}
