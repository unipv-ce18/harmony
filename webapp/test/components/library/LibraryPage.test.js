import 'regenerator-runtime/runtime';   // for Babel to polyfill "async"

import {mount} from 'enzyme';
import waitForExpect from 'wait-for-expect';

import '../../matchMedia.mock';

import LibraryPage from '../../../src/components/library/LibraryPage';
import {getLibrary, getUserInfo, getUserPlaylists} from '../../../src/core/apiCalls';


jest.mock('../../../src/core/Session', () => ({
  Session: jest.fn().mockImplementation(() => ({
      getAccessToken: jest.fn().mockResolvedValue('MOCK_ACCESS_TOKEN'),
      addStatusListener: jest.fn()
  }))
}));
jest.mock('../../../src/core/apiCalls');
getUserInfo.mockResolvedValue({});
getUserPlaylists.mockResolvedValue([{id: 'aaa', name: 'aaa'}]);

getLibrary.mockResolvedValue(require('./test-library.json'));


describe('LibraryPage', () => {
  beforeEach(() => {
    getLibrary.mockClear();
  });

  it('should display the user\'s library', async () => {
    const wrapper = mount(<LibraryPage id="me"/>)

    // Expands all the sections
    wrapper.setState({
      artists: true,
      playlists: true,
      releases: true,
      songs: true
    });

    await waitForExpect(() => {
      expect(getLibrary).toHaveBeenCalledTimes(1);
      expect(getLibrary).toHaveBeenCalledWith('me', 'MOCK_ACCESS_TOKEN', true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
