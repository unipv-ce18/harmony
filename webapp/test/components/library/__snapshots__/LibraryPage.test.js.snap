// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibraryPage should display the user's library 1`] = `"<div class=\\"libraryPage\\"><h1>LIBRARY</h1><div><div class=\\"libraryMenues\\"><span class=\\"iconButton\\" title=\\"Collapse Playlist\\" style=\\"width: 20px; height: 20px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span><span>PLAYLISTS</span><span class=\\"iconButton\\" title=\\"Collapse Playlist\\" style=\\"width: 20px; height: 20px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span></div><div><div><div class=\\"libraryCommon\\"><div><hr><p>Realized by you</p></div><span><a href=\\"#\\"><img src=\\"test-file-stub\\" alt=\\"\\"></a><p><a href=\\"#\\">New Playlist</a></p></span><span><a href=\\"#\\"><div class=\\"divImage singleimage\\" style=\\"width: 150px; height: 150px;\\"><div><img src=\\"test-file-stub\\" alt=\\"\\"></div></div></a><p><a href=\\"#\\">Saem Playlist</a></p></span></div><div class=\\"libraryCommon\\"></div></div></div><div class=\\"libraryMenues\\"><span class=\\"iconButton\\" title=\\"Collapse Artists\\" style=\\"width: 20px; height: 20px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span><span>ARTISTS</span><span class=\\"iconButton\\" title=\\"Collapse Artists\\" style=\\"width: 20px; height: 20px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span></div><div><div class=\\"libraryArtists\\"><span><a href=\\"#\\"><img src=\\"test-file-stub\\" alt=\\"\\"></a><p><a href=\\"#\\">Linkin Park</a></p></span></div></div><div class=\\"libraryMenues\\"><span class=\\"iconButton\\" title=\\"Collapse Releases\\" style=\\"width: 20px; height: 20px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span><span>RELEASES</span><span class=\\"iconButton\\" title=\\"Collapse Releases\\" style=\\"width: 20px; height: 20px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span></div><div><div class=\\"libraryCommon\\"><span><a href=\\"#\\"><img src=\\"test-file-stub\\" alt=\\"\\"></a><p><a href=\\"#\\">Meteora</a></p><p>By <a href=\\"#\\">Linkin Park</a></p></span></div></div><div class=\\"libraryMenues\\"><span class=\\"iconButton\\" title=\\"Collapse Songs\\" style=\\"width: 20px; height: 20px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span><span>SONGS</span><span class=\\"iconButton\\" title=\\"Collapse Songs\\" style=\\"width: 20px; height: 20px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span></div><div><div class=\\"songsInfo\\"><table><tr><th></th><th></th><th><button>Title</button></th><th><button>Artist</button></th><th><button>Release</button></th><th><button>Time</button></th><th></th><th><button>Pitch</button></th><th></th><th><button>Plays</button></th></tr><tr><td class=\\"hidenButtons\\"><span class=\\"iconButton\\" title=\\"Play\\" style=\\"width: 22px; height: 22px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span></td><td><span class=\\"iconButton\\" title=\\"Like\\" style=\\"width: 24px; height: 24px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span></td><td><p>Foreword</p></td><td><a href=\\"#\\">Linkin Park</a></td><td><a href=\\"#\\">Meteora</a></td><td>0:20</td><td></td><td></td><td></td><td></td><td class=\\"hidenButtons\\"><span class=\\"iconButton\\" title=\\"Menu\\" style=\\"width: 24px; height: 24px;\\"><svg><use xlink:href=\\"undefined\\"></use></svg></span></td></tr></table></div></div></div></div>"`;
